shedlock-provider-hazelcast/src/main/java/net/javacrumbs/shedlock/provider/hazelcast/HazelcastLockProvider.java

```diff
@@ -0,0 +1,191 @@
+            removeLock(lockName);
+            addNewLock(lockConfiguration);
+            return true;
+        } else {
+            log.debug("lock - already locked : currentLock={};  conf={}", lock, lockConfiguration);
+            return false;
+        }
+    }
+
+    private IMap<String, HazelcastLock> getStore() {
+        return hazelcastInstance.getMap(lockStoreKey);
+    }
+
+    protected HazelcastLock getLock(final String lockName) {
+        final IMap<String, HazelcastLock> store = getStore();
+        return store.get(lockName);
+    }
+
+    private void removeLock(final String lockName) {
+        final IMap<String, HazelcastLock> store = getStore();
+        store.delete(lockName);
+        log.debug("lock store - lock deleted : {}", lockName);
+    }
+
+    private void addNewLock(final LockConfiguration lockConfiguration) {
+        final String localMemberUuid = getLocalMemberUuid();
+        final HazelcastLock lock = HazelcastLock.fromLockConfiguration(lockConfiguration, localMemberUuid);
+        final String lockName = lockConfiguration.getName();
+        final IMap<String, HazelcastLock> store = getStore();
+        store.put(lockName, lock);
+        log.debug("lock store - new lock added : {}", lock);
+    }
+
+    private String getLocalMemberUuid() {
+        return hazelcastInstance.getCluster().getLocalMember().getUuid();
+    }
+
+    protected boolean isUnlocked(final HazelcastLock lock) {
+        return lock == null;
+    }
+
+    protected boolean isExpired(final HazelcastLock lock, final Instant now) {
+        final Instant unlockTime = lock.getUnlockTime();
+        return now.isAfter(unlockTime);
+    }
+
+    protected boolean isLockedByUnvailableMemberOfCluster(final HazelcastLock lock) {
+        final String memberUuid = lock.getClusterMemberUuid();
+        final boolean membreIsUp = hazelcastInstance.getCluster().getMembers().stream().anyMatch(member -> member.getUuid().equals(memberUuid));
+        return !membreIsUp;
+    }
+
+    /**
+     * Unlock the lock with its name.
+     *
+     * @param lockName the name of the lock to unlock.
+     */
+    protected void unlock(final String lockName) {
+        log.trace("unlock - attempt : {}", lockName);
+        final Instant now = Instant.now();
+        final IMap<String, HazelcastLock> store = getStore();
+        try {
+            store.lock(lockName);
+            final HazelcastLock lock = getLock(lockName);
+            unlockProperly(lock, now);
+        } finally {
+            store.unlock(lockName);
+        }
+    }
+
+    private void unlockProperly(final HazelcastLock lock, final Instant now) {
+        if (isUnlocked(lock)) {
+            log.debug("unlock - it is already unlocked");
+            return;
+        }
+        final String lockName = lock.getName();
+        final Instant lockAtLeastInstant = lock.getLockAtLeastUntil();
+        if (now.isAfter(lockAtLeastInstant)) {
```

Comment:
It would have been better to change the condition to !now.isBefore(). shouldBeAbleToLockRightAfterUnlock test is sometimes fails if lockAtLeastFor = 0 and we manage to explicitly unlock and attempt to acquire the lock in the same millisecond (my computer is really fast).